### DISCUSSION TITLE
Dagre auto-layout not working properly after editing the layout ( delete node, and insert new node in-between nodes )

### DISCUSSION THREAD

**Discussion by dexkode4** [xyflow/xyflow#4251] [Layouting]
### Describe the Bug

I'm building a binary tree structure node visualisation web-app
I want to be able to re-render all nodes with the new position after deleting a node

**Current behaviour**
On adding a new node after deleting a node, the new node jumps to a new position as the deleted node is still present

**code sample**

```
const CustomNodeFlow = () => {
  const dispatch = useAppDispatch();
  const { elements } = useAppSelector((state) => state.flowgraphSlice);

  const layeredElements = useMemo(
    () => getLayoutedElements(elements, 'LR'),
    [elements]
  );

  const onElementClick = (
    event: React.MouseEvent<Element, MouseEvent>,
    element: Node | Edge
  ) => {
    if (element.type === NodeTypesEnum.DecisionNode) return;
    dispatch(setSelectedNode(element as Node));
  };

  return (
    <ReactFlow
      elements={layeredElements}
      nodeTypes={nodeTypes}
      onElementClick={onElementClick}
      defaultZoom={0.75}
    >
      <Background variant={BackgroundVariant.Dots} gap={24} size={0.5} />
      <Controls />
    </ReactFlow>
  );
};
```

### Steps to Reproduce the Bug or Issue

On adding a new node after deleting a node, the new node jumps to a new position as the deleted node is still present

### Expected behavior

The graph should re-render  nodes with newly calculated position

### Screenshots or Videos

https://www.loom.com/share/e918c73a3e534e1f8c7fb72b159fff16

### Platform

- OS: macOS
- Browser: Chrome, Safari, Firefox
- Version: 10

**Comment by raarts** [comment]
My $ 0.02: 

- please provide the getLayoutedElements() function. 
- How do you node the other node is still present?
- Do you recreate the Dagre Graph from scratch every time you layout?

**Comment by moklick** [comment]
Please update to the latest version v.10.0.4

**Comment by dexkode4** [comment]
> My $ 0.02:
> 
> * please provide the getLayoutedElements() function.
> * How do you node the other node is still present?
> * Do you recreate the Dagre Graph from scratch every time you layout?
@raarts  thanks for the reply
```
const defaultNodeWidth = 252;
const defaultNodeHeight = 200;

export const getLayoutedElements = (
  elements: Elements,
  direction: 'TB' | 'LR'
) => {
  const isHorizontal = direction === 'LR';
  dagreGraph.setGraph({ rankdir: direction });

  dagreGraph.setDefaultEdgeLabel(() => ({}));
  elements?.forEach((el) => {
    if (isNode(el)) {
      dagreGraph.setNode(el.id, {
        label: el.id,
         width: el?.__rf?.width || defaultNodeWidth,
         height: el?.__rf?.height || defaultNodeHeight,
         });
    } else {
      dagreGraph.setEdge(el.source, el.target);
    }
  });

  dagre.layout(dagreGraph);

  return elements.map((el) => {
    const nodeWithPosition = dagreGraph.node(el.id);
    return {
      ...el,
      targetPosition: isHorizontal ? Position.Top : Position.Top,
      sourcePosition: isHorizontal ? Position.Top : Position.Top,
      position : {
        x: 50 + nodeWithPosition?.x - nodeWithPosition?.width / 2 + Math.random() / 1000,
        y: nodeWithPosition?.y - nodeWithPosition?.height / 2 + 50,
      }
    };

  });
};

```

3rd question
I add/remove nodes to/from a redux store state `elements`

on each change to the `elements`, I call `getLayoutedElements()`

**Comment by raarts** [comment]
Try starting getLayoutedElements with this:

```
  const dagreGraph = new dagre.graphlib.Graph();  // <----- watch this line
  dagreGraph.setDefaultEdgeLabel(() => ({}));
  const isHorizontal = direction === 'LR';
  dagreGraph.setGraph({ rankdir: direction });
```

**Comment by dexkode4** [comment]
> Try starting getLayoutedElements with this:
> 
> ```
>   const dagreGraph = new dagre.graphlib.Graph();  // <----- watch this line
>   dagreGraph.setDefaultEdgeLabel(() => ({}));
>   const isHorizontal = direction === 'LR';
>   dagreGraph.setGraph({ rankdir: direction });
> ```

Thank you so much for this, it worked for deleting a node
but it didn't fix all the issues ( sorry I didn't mention earlier ), please check this clip
https://www.loom.com/share/41ce9d3096344f6aaabfc4b8fc0ea6c4

**Issue**
When I insert a node in-between existing node, in that order from the clip, the layout gets weird

Expectation
Green branch stays Top always

**Comment by raarts** [comment]
Seen this happen in my setup too, but it wasn't an issue for me. 
My guess: you will have to study dagre internals if you want to sort this out.

**Comment by jayanthbondi** [comment]
Hi @dexkode4 , I'm facing the similar issue, could you please share the codesandbox link if you have it.
